### DISCUSSION TITLE
Static generation server side dynamic import

### DISCUSSION THREAD

**Discussion by Cybertron1** [vercel/next.js#31944] [Help]
Hi everyone,
I am currently building a CMS which should create statically generated pages.
Since it is a simple CMS the content of the pages is stored in a JSON. This means I have a [...path].js page which will then build all the pages during compilation.
Now here is my problem.
Since I can't know what the content will be, I want to dynamically load the blocks (a block is just a React Component with a default export) into the page. Since I can't pass react elements from getStaticProps into the component I have to do it in the component itself.
But dynamic imports are asynchronous, which means I have to wrap it into an useEffect which won't be executed on the server which breaks static generation.
How can I do this? Did I miss something?

Here is some code to hopefully make it understandable:

templates/template.json
```json
[
  {
    "type": "text"
  }
]
```

blocks/text.jsx
```javascript
const Text = () => {
return <div>text</div>
};
export default Text;
```

lib.js
```javascript
async function fetchTemplate() {
  const template = path.join(process.cwd(), 'templates', `template.json`);
  const fileContent = await fs.readFile(template, 'utf-8')
  return JSON.parse(fileContent);
}
```

[...path].jsx
```javascript
export async function getStaticPaths() {
  return {
    paths: [
      { params: { path: ['a', 'b'] }
    ],
    fallback: false
  };
}
export async function getStaticProps({ params }) {
  const template = await fetchTemplate();

  const blocks = await Promise.all(template.map(async block => {
    const blockData = await import(`../blocks/${block.type}.jsx`);

    return blockData.default;
  }));

  // what to do with blocks -> how can I do this to achieve static generation

  return {
    props: {
      template
      // blocks -> won't work
    }
  };
}

export default function Page({ template }) {
  return <div></div>;
}
```

**Comment by tpetersons** [comment]
Hi,
Isn't there dynamic import from next JS which supports SSR?
https://nextjs.org/docs/advanced-features/dynamic-import
Also just a thought - how about using "require" instead of "import"? I believe it's not async.
